lte/gateway/c/session_manager/SessionCredit.cpp

```diff
@@ -304,6 +304,49 @@ SessionCredit::Usage SessionCredit::get_usage_for_reporting(
   return usage;
 }
 
+RequestedUnits SessionCredit::get_requested_credits_units() {
+  RequestedUnits requestedUnits;
+  uint64_t buckets_used_total =buckets_[USED_TX] + buckets_[USED_RX];
+
+  uint64_t total_requested = calculate_requested_unit(
+      received_granted_units_.total(),
+      ALLOWED_TOTAL, ALLOWED_FLOOR_TOTAL, buckets_used_total);
+
+  uint64_t tx_requested = calculate_requested_unit(
+      received_granted_units_.tx(),
+      ALLOWED_TX, ALLOWED_FLOOR_TX, buckets_[USED_TX]);
+
+  uint64_t rx_requested = calculate_requested_unit(
+      received_granted_units_.rx(),
+      ALLOWED_RX, ALLOWED_FLOOR_RX, buckets_[USED_RX]);
+
+  requestedUnits.set_total(total_requested);
+  requestedUnits.set_tx(tx_requested);
+  requestedUnits.set_rx(rx_requested);
+
+  return requestedUnits;
+}
+
+// returns either the last grant, or the difference between the last grant
+// and the credit remaining. Prevents over requesting in case we still have
+// credit available from the previous request
+uint64_t SessionCredit:: calculate_requested_unit(
+    CreditUnit cu, Bucket allowed,Bucket allowed_floor, uint64_t used){
+  if (cu.is_valid() == false){
+    return 0;
+  }
```

Comment:
How do we end up in a scenario when `cu.volumt()` is 0 but `buckets_[allowed]-buckets_[allowed_floor]` is not?

Reply:
`cu` comes from the last Grant. So for example, if you requested 0 because you had enough, and OCS gave you 0, then you will save this 0 as a grant (which will not work in this calculation)

However, when you get 0 from the gsu, the Allowed_Floor does not advance. So you can reconstruct the grant from `Allowed - Allowed_floor`

So I think we may be able to use just `Allowed - Allowed_floor` for this calculation

Reply:
Ohh I see, so you are calculating the most recent non-0 grant?

Reply:
Now I remember why we are using both:
Case 1: regular case, we get a valid grant, and we don't overreport. Then we can use either last grant or Allowed - Allowed_floor to reconstruct the last valid grant.

Case 2: (what i mentioned before) if grant is 0, we need to reconstruct last non 0 grant received to be able to request something. For 

Case 3: in case of going over the grant, the ALLOWED will be raised automatically by the algoritm. So if we were using `Allowed - Allowed_floor` to calculate the last grant received, that grant would grow and we would be requesting more every time. 

Note that when we go over `Allowed` limit but it is not final unit, we will not terminate. So we consider that the next time PCRF/OCS answers will take into consideration that over reporting and respond with a grant.

Reply:
Got it, thank you for the explanation!